### PR TITLE
[otbn,dv] Add the "BadAtEnd" generator to RIG

### PR DESCRIPTION
**Pull request by rswarbrick**
The main commit is the last one in the series: the other two are refactorings to make that easier. The message for the last commit is:

> This generates a branch, jump or loop as the last instruction in a
> loop.
> 
> To do this, we also factor out the gen_head method in the Branch
> generator. This is the bit that generates the actual branch
> instruction, which is what we need for BadAtEnd.
> 
> Similarly, we factor out a _pick_head method in the Loop generator.
> This is in charge of picking the LOOP/LOOPI instruction (it also
> returns some auxiliary information about loop shape that we need).
>

### CODE REVIEW

File: hw/ip/otbn/dv/rig/rig/configs/base.yml [hunk 1]
```diff
@@ -12,6 +12,7 @@ gen-weights:
 
   # Generators that end the program
   ECall: 1
+  BadAtEnd: 5
```
Comment (imphil) [inline hw/ip/otbn/dv/rig/rig/configs/base.yml:15]: Are you intentionally giving this generator such a high weight, or is this a leftover from testing?

File: hw/ip/otbn/dv/rig/rig/gens/bad_at_end.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+        # we also have to teleport old_model into the right place!
+        old_model.pc = model.pc
+        model.merge(old_model)
+
+        # Pick a new instruction to replace the last one we generated. Start by
+        # shuffling the numbers from 0..2, which will represent Branch, Loop,
+        # Jump respectively. The idea is that we want to try them in an
+        # arbitrary order, but don't want to fail completely if one of the
+        # other generators might have worked.
+        types = list(range(2))
```
Comment (imphil) [inline hw/ip/otbn/dv/rig/rig/gens/bad_at_end.py:72]: Should be `range(3)` most likely?

**Review by imphil — APPROVED** [review]
LGTM with probably one bug to be addressed before merging.

File: hw/ip/otbn/dv/rig/rig/gens/bad_at_end.py (hunk 2 above)
Comment (rswarbrick) [reply hw/ip/otbn/dv/rig/rig/gens/bad_at_end.py:72]: Ah, that explains why some of the coverage points weren't being hit much :-)

File: hw/ip/otbn/dv/rig/rig/configs/base.yml (hunk 1 above)
Comment (rswarbrick) [reply hw/ip/otbn/dv/rig/rig/configs/base.yml]: Oops, thanks. Fixed.

**Pull request merged**